Loop/Managers/Missed Meal Detection/MealDetectionManager.swift

```diff
@@ -0,0 +1,311 @@
+//
+//  MealDetectionManager.swift
+//  Loop
+//
+//  Created by Anna Quinlan on 11/28/22.
+//  Copyright © 2022 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+import HealthKit
+import OSLog
+import LoopCore
+import LoopKit
+
+enum MissedMealStatus: Equatable {
+    case hasMissedMeal(startTime: Date, carbAmount: Double)
+    case noMissedMeal
+}
+
+class MealDetectionManager {
+    private let log = OSLog(category: "MealDetectionManager")
+    
+    public var maximumBolus: Double?
+    
+    /// The last missed meal notification that was sent
+    /// Internal for unit testing
+    var lastMissedMealNotification: MissedMealNotification? = UserDefaults.standard.lastMissedMealNotification {
+        didSet {
+            UserDefaults.standard.lastMissedMealNotification = lastMissedMealNotification
+        }
+    }
+    
+    private var carbStore: CarbStoreProtocol
+    
+    /// Debug info for missed meal detection
+    /// Timeline from the most recent check for missed meals
+    private var lastEvaluatedMissedMealTimeline: [(date: Date, unexpectedDeviation: Double?, mealThreshold: Double?, rateOfChangeThreshold: Double?)] = []
+    
+    /// Timeline from the most recent detection of an missed meal
+    private var lastDetectedMissedMealTimeline: [(date: Date, unexpectedDeviation: Double?, mealThreshold: Double?, rateOfChangeThreshold: Double?)] = []
+    
+    /// Allows for controlling uses of the system date in unit testing
+    internal var test_currentDate: Date?
+    
+    /// Current date. Will return the unit-test configured date if set, or the current date otherwise.
+    internal var currentDate: Date {
+        test_currentDate ?? Date()
+    }
+
+    internal func currentDate(timeIntervalSinceNow: TimeInterval = 0) -> Date {
+        return currentDate.addingTimeInterval(timeIntervalSinceNow)
+    }
+    
+    public init(
+        carbStore: CarbStoreProtocol,
+        maximumBolus: Double?,
+        test_currentDate: Date? = nil
+    ) {
+        self.carbStore = carbStore
+        self.maximumBolus = maximumBolus
+        self.test_currentDate = test_currentDate
+    }
+    
+    // MARK: Meal Detection
+    func hasMissedMeal(insulinCounteractionEffects: [GlucoseEffectVelocity], completion: @escaping (MissedMealStatus) -> Void) {
+        let delta = TimeInterval(minutes: 5)
+
+        let intervalStart = currentDate(timeIntervalSinceNow: -MissedMealSettings.maxRecency)
+        let intervalEnd = currentDate(timeIntervalSinceNow: -MissedMealSettings.minRecency)
+        let now = self.currentDate
+
+        carbStore.getGlucoseEffects(start: intervalStart, end: now, effectVelocities: insulinCounteractionEffects) {[weak self] result in
```

Comment:
It might be a nice separation of concerns to pass in carbEffects, rather than asking carbstore for it, and creating a dependency on carbstore for this detector.

Reply:
I agree it would be good to remove this dependency. 

My only hesitation is that the `carbEffects` that are computed via `update()` in `LoopDataManager` go back ~30 minutes ("last BG time - retrospective correction interval"). Since the missed meal detection wants carb effects that go back 2 hours (`MissedMealSettings.maxRecency`), this would mean needing to call `getCarbEffects` again at the `LoopDataManager` level. This `getCarbEffects` call would need to know the `MissedMealSettings.maxRecency`, which feels like implementation details leaking out of the class. What do you think would be best?

Reply:
I think the call again at LDM level is fine. As far as knowing how much data the MissedMeal algorithm needs, and what is considered "implementation details" it depends on if you want maxRecency to be an algorithm configuration option (in which case it is public, and the api documentation should state that the carb effects passed in should cover a range matching the configuration option), or make it a static amount (in which case the docs specify the needed time range that the passed in carbs need to cover).

Reply:
Yeah, I was hoping to make it a configuration option at build-time. I'll update it to use carb effects and make sure the documentation states this assumption